src/parallel/MainSplitter.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2022, Antti Hyvarinen <antti.hyvarinen@gmail.com>
+ * Copyright (c) 2022, Seyedmasoud Asadzadeh <seyedmasoud.asadzadeh@usi.ch>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#ifndef OPENSMT_MAINSPLITTER_HPP
+#define OPENSMT_MAINSPLITTER_HPP
+
+#include "MainSolver.h"
+#include "SplitData.h"
+#include "ScatterSplitter.h"
+#include "OsmtInternalException.h"
+#include "LookaheadSplitter.h"
+#include "Splitter.h"
+
+#include <cmath>
+
+class MainSplitter : public MainSolver {
+public:
+
+    bool isSplitTypeScatter() const & { return dynamic_cast<Splitter&>(ts.solver).isSplitTypeScatter(); }
+
+    bool isSplitTypeNone() const & { return dynamic_cast<Splitter&>(ts.solver).isSplitTypeNone(); }
+
+    MainSplitter(std::unique_ptr<Theory> t,std::unique_ptr<TermMapper> tm, std::unique_ptr<THandler> th,
+                 std::unique_ptr<SimpSMTSolver> ss, Logic & logic, SMTConfig & config, std::string name)
+                 :
+                 MainSolver(std::move(t), std::move(tm), std::move(th), std::move(ss),logic,config, std::move(name))
+                 {}
+
+    PTPLib::net::Channel & getChannel()  { return (dynamic_cast<ScatterSplitter&>(getSMTSolver())).getChannel(); }
+
+    void notifyResult(sstat const & result)
+    {
+        if (result != s_Undef) {
+            getChannel().setShallStop();
+            getChannel().notify_all();
+        }
+    }
+
+    sstat check() override {
+        if (not config.sat_solver_limit())
+        //push frames size should match with length of the solver branch
+            if (frames.size() != static_cast<std::size_t>((dynamic_cast<ScatterSplitter&>(getSMTSolver())).get_solver_branch().size() + 1))
+                throw PTPLib::common::Exception(__FILE__, __LINE__,";assert: Inconsistency in push frames size and length of the solver address");
+
+        sstat res = MainSolver::check();
+        notifyResult(res);
+        status = res;
+        return res;
+    }
+
+    sstat solve_(vec<FrameId> & enabledFrames) override {
+        if (not config.sat_solver_limit()) {
```

Comment:
Since we have several usage of "dynamic_cast<ScatterSplitter&>(getSMTSolver())", how about following ?
```
ScatterSplitter & getScatterSplitter()  { return dynamic_cast<ScatterSplitter&>(getSMTSolver()) ;}
```

Reply:
Sounds good!